syntax = "proto3";

package synapse;

import "google/protobuf/empty.proto";
import "api/node.proto";
import "api/query.proto";
import "api/status.proto";
import "api/files.proto";
import "api/logging.proto";

message Peripheral {
  enum Type {
    kUnknown = 0;
    kBroadbandSource = 1;
    kElectricalStimulation = 2;
    kOpticalStimulation = 3;
    kSpikeSource = 4;
  }
  string name = 1;
  string vendor = 2;
  uint32 peripheral_id = 3;
  Type type = 4;
  string address = 5;
}

message DeviceInfo {
  string name = 1;
  string serial = 2;
  uint32 synapse_version = 3;
  uint32 firmware_version = 4;
  Status status = 5;
  repeated Peripheral peripherals = 6;
  DeviceConfiguration configuration = 7;
}

message DeviceConfiguration {
  repeated NodeConfig nodes = 1;
  repeated NodeConnection connections = 2;
}

message TapDownsampleOptions {
  
}

message CreateTapRequest {
  // Which node to tap the output from
  uint32 node_id = 1;

  // Client address to the send data to
  string client_address = 2;

  // Client port to send data to
  uint32 client_port = 3;

  // Downsample options specific to this tap
  TapDownsampleOptions downsample_options = 4;
}

message CreateTapResponse {
  StatusCode code = 1;
}

message RemoveTapRequest {
  uint32 node_id = 1;
  string client_address = 2;
  uint32 client_port = 3;
}

message RemoveTapResponse {
  StatusCode code = 1;
}


service SynapseDevice {
  rpc Info(google.protobuf.Empty) returns (DeviceInfo) {}
  rpc Configure(DeviceConfiguration) returns (Status) {}
  rpc Start(google.protobuf.Empty) returns (Status) {}
  rpc Stop(google.protobuf.Empty) returns (Status) {}
  rpc Query(QueryRequest) returns (QueryResponse) {}
  rpc StreamQuery(StreamQueryRequest) returns (stream StreamQueryResponse) {}

  rpc ListFiles(google.protobuf.Empty) returns (ListFilesResponse) {}
  rpc WriteFile(WriteFileRequest) returns (WriteFileResponse) {}
  rpc ReadFile(ReadFileRequest) returns (stream ReadFileResponse) {}
  rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse) {}
  rpc GetLogs(LogQueryRequest) returns (LogQueryResponse) {}
  rpc TailLogs(TailLogsRequest) returns (stream LogEntry) {}

  rpc CreateTap(CreateTapRequest) returns (CreateTapResponse) {}
  rpc RemoveTap(RemoveTapRequest) returns (RemoveTapResponse) {}
}
